code/modules/reagents/chemistry/machinery/smoke_machine.dm

```diff
@@ -0,0 +1,139 @@
+/obj/machinery/smoke_machine
+	name = "Smoke Machine"
+	icon = 'icons/obj/chemical.dmi'
+	icon_state = "smoke0"
+	density = TRUE
+	anchored = TRUE
+	circuit = /obj/item/circuitboard/machine/smoke_machine
+	var/efficiency = 10
+	var/on = FALSE
+	var/cooldown = 0
+	var/screen = "home"
+	var/analyzeVars[0]
+	var/useramount = 30 // Last used amount
+	var/volume = 1000
+	var/setting = 3
+	var/list/possible_settings = list(3,6,9,12,15)
+
+/datum/effect_system/smoke_spread/chem/smoke_machine/set_up(datum/reagents/carry = null, setting = 3, efficiency = 10, loc)
+	amount = setting
+	carry.copy_to(chemholder, 20)
+	carry.remove_any(setting * 16 / efficiency)
+	location = loc
+
+/obj/machinery/smoke_machine/Initialize()
+	create_reagents(volume)
+	. = ..()
+/obj/machinery/smoke_machine/Destroy()
+	STOP_PROCESSING(SSprocessing, src)
+	return ..()
+
+/obj/machinery/smoke_machine/update_icon()
+	if((stat & NOPOWER) || (!on) || (reagents.total_volume == 0))
+		icon_state = "smoke0"
+	else
+		icon_state = "smoke1"
+	. = ..()
+
+
+/obj/machinery/smoke_machine/RefreshParts()
+	efficiency = 6
+	for(var/obj/item/stock_parts/matter_bin/B in component_parts)
+		efficiency += B.rating
+	for(var/obj/item/stock_parts/capacitor/C in component_parts)
+		efficiency += C.rating
+	for(var/obj/item/stock_parts/manipulator/M in component_parts)
+		efficiency += M.rating
+
+/obj/machinery/smoke_machine/process()
+	..()
+	update_icon()
+	if(stat & NOPOWER)
+		return
+	if(reagents.total_volume == 0)
+		on = FALSE
+		return
+	var/turf/T = get_turf(src)
+	var/smoke_test
+	smoke_test = locate(/obj/effect/particle_effect/smoke) in T.contents
+	if(on && (!smoke_test))
+		var/datum/effect_system/smoke_spread/chem/smoke_machine/smoke = new()
+		smoke.set_up(reagents, setting, efficiency, get_turf(src))
+		smoke.start()
+
+/obj/machinery/smoke_machine/attackby(obj/item/I, mob/user, params)
+	add_fingerprint(user)
+	if(istype(I, /obj/item/reagent_containers))
+		var/obj/item/reagent_containers/RC = I
+		var/units = RC.reagents.trans_to(src, RC.amount_per_transfer_from_this)
```

Comment:
would help to add logging for adding reagents tbh but w/e if you dont im not going to throw a bitchfit lol.

Reply:
That's excessuve

Reply:
Yea, my experience is that sort of thing isn't helpful. The loaded reagents are clearly listed and so final responsibility should always rest with whoever flips the "on" switch.